klippy/mcu.py

```diff
@@ -230,27 +230,31 @@ def _build_config(self):
             return
         self._oid = self._mcu.create_oid()
         self._mcu.add_config_cmd(
-            "config_soft_pwm_out oid=%d pin=%s cycle_ticks=%d value=%d"
+            "config_soft_pwm_out oid=%d pin=%s value=%d"
             " default_value=%d max_duration=%d"
-            % (self._oid, self._pin, cycle_ticks,
-               self._start_value >= 1.0, self._shutdown_value >= 0.5,
-               self._mcu.seconds_to_clock(self._max_duration)))
+            % (self._oid, self._pin, self._start_value >= 1.0,
+                self._shutdown_value >= 0.5,
+                self._mcu.seconds_to_clock(self._max_duration)))
         curtime = self._mcu.get_printer().get_reactor().monotonic()
         printtime = self._mcu.estimated_print_time(curtime)
         self._last_clock = self._mcu.print_time_to_clock(printtime + 0.100)
         svalue = int(self._start_value * self._pwm_max + 0.5)
         self._mcu.add_config_cmd(
-            "schedule_soft_pwm_out oid=%d clock=%d on_ticks=%d"
-            % (self._oid, self._last_clock, svalue), is_init=True)
+            "schedule_soft_pwm_out oid=%d clock=%d on_ticks=%d off_ticks=%d"
+            % (self._oid, self._last_clock, svalue, cycle_ticks - svalue),
+                is_init=True)
         self._set_cmd = self._mcu.lookup_command(
-            "schedule_soft_pwm_out oid=%c clock=%u on_ticks=%u",
+            "schedule_soft_pwm_out oid=%c clock=%u on_ticks=%u off_ticks=%u",
             cq=cmd_queue)
-    def set_pwm(self, print_time, value):
+    def set_pwm(self, print_time, value, cycle_time=None):
```

Comment:
If you add a new field to the set_pwm() method, you will also need to update all the other set_pwm() implementations in Klipper (MCU_digital, sx1509, multi_pin, etc.).

Reply:
Is it really necessary? The new argument has a default value, so the behavior should be the same if it is not supplied.

Reply:
Yes, because one may define an [output_pin] config section that uses an sx1509 pin, a multi_pin, etc.

-Kevin

Reply:
I added the new argument. Let me know if I missed it somewhere.
The cycle time seems to be unused in the sx1509 module.